src/vs/workbench/services/extensions/common/abstractExtensionService.ts

```diff
@@ -803,29 +804,46 @@ export class ExtensionRunningLocationClassifier {
 
 class ProposedApiController {
 
-	private readonly enableProposedApiFor: string[];
+	private readonly enableProposedApiFor: Promise<string[]>;
 	private readonly enableProposedApiForAll: boolean;
 	private readonly productAllowProposedApi: Set<string>;
 
 	constructor(
 		@IWorkbenchEnvironmentService private readonly _environmentService: IWorkbenchEnvironmentService,
-		@IProductService productService: IProductService
+		@IProductService productService: IProductService,
+		@IFileService fileService: IFileService
 	) {
+
 		// Make enabled proposed API be lowercase for case insensitive comparison
-		this.enableProposedApiFor = (_environmentService.extensionEnabledProposedApi || []).map(id => id.toLowerCase());
+		const cliProposedApi = (_environmentService.extensionEnabledProposedApi || []).map(id => id.toLowerCase());
+
+		let enableProposedAPIResolver: (ids: string[]) => void;
+		this.enableProposedApiFor = new Promise(c => enableProposedAPIResolver = c);
+
+		runWhenIdle(async () => {
+			const argvContent = await fileService.readFile(_environmentService.argvResource);
+			const argvString = argvContent.value.toString();
```

Comment:
I am confused. So, we do have support for an `argv`-file but this isn't automatically picked up by the enviroment service and applied to [`INativeEnvironmentService#args`](https://github.com/microsoft/vscode/blob/029ba8994753a882a81a6986f7f0754691e891be/src/vs/platform/environment/common/environment.ts#L107) but every potential consumer needs to read now two sources of truth? @bpasero Is this correct?

Reply:
Sorry, but I will take myself out of this discussion. The original intent was to allow to configure Electron command line arguments that have to be set very very early before `ready` event. Then someone (not me) added support for enabling proposed API. I was against this from the beginning iirc and I don't know what to fix here. 

Bottom line: `argv.json` is NOT for adding something to environment service. It is for configuring things as early as possible.

I would suggest to move this into a setting.

Reply:
I'd prefer for it to be in the Environment service, and that's what the original implementation was trying to do, but the old implementation stopped working recently as it seems modifications to process.argv here https://github.com/microsoft/vscode/pull/110461/files#diff-27653c212e1cfe533e4eb2f7d0d3f89604c9de48a09583b4cbbbcbd08a07da79L265 were not getting picked up here https://github.com/microsoft/vscode/blob/3ca69614161b395ac80ce747c483f623743b4f52/src%2Fvs%2Fworkbench%2Fservices%2Fenvironment%2Felectron-browser%2FenvironmentService.ts#L75. Additionally extensionEnabledProposedApi needs to be available synchronously so I wasn't sure how it to make it read a file there. Any tips for making this cleaner would be appreciated.

Reply:
> I don't know what to fix here.

I'd say that's easy: whatever is defined in the `argv`-file must overwrite what's exposed via `argv`-attribute of the environment service.

Reply:
The problem there is the environment service is initialized very early and blocks the rest of the startup process. So I don't want to wait for a file read as it'd slow down startup.

Reply:
I believe we [already block startup](https://github.com/microsoft/vscode/blob/9eedabce7b8111258fa734689566237730156871/src/main.js#L236) because of reading this file. So, I'd be favour of reading it only once and never again. Also, it merging with the "other"  arguments seems elegant to me because a down-level consumer shouldn't really care about the existence of this file.

Reply:
True, we do read it there. I guess the main issue is I'm not sure how I can communicate between whatever runtime is running that file and the main process. In the past simply appending to `process.argv` worked, but that seems to be growing increasingly flakey.

Reply:
> I'm not sure how I can communicate between whatever runtime is running that file and the main process.

Sounds like a challenge to me

Reply:
I agree that it is not good that we read directly the `argv.json` file from the extension service. The `ProposedApiController` is instantiated on the web too and I find it very weird that we would try to read a file on the web. IMHO this should come through a service, either `IEnvironmentService` or `IConfigurationService`. Adding a competing configuration pattern that is only used here is IMHO not so good.